js/rank_states_utils.js

```diff
@@ -112,84 +142,170 @@ function rankEm() {
-            .style('stroke', 'rgb(200, 65, 39)') // emphasize with a redorange stroke
-            .transition().duration(600)
-              .style('stroke','rgb(190,190,190)')
-              .style('stroke-dashoffset',"12"); // needs to be a multiple of stroke-dasharray to look smooth
-          barchoice.style('stroke-dashoffset',null); // resets to "0", which doesn't change the look
-          d3.select(this)
-          .transition().duration(600)
-            .attr('transform','translate(0,0)');  
-        }
-      }
-      
+rankSvg.updateStyles = function(){
+  var allBars = svgStates.select('#ranked-states-bars').selectAll('rect');
+  var lockedStates = svgStates.select('#ranked-states-locked').selectAll('.locked-state');
+  var draggableStates = svgStates.select('#ranked-states-draggable').selectAll('.draggable-state');
+  var lockedBars = allBars.filter('.locked-rank-bar');
+  var openBars = allBars.filter('*:not(.locked-rank-bar)');
+  
+  function clearHighlight(){
+    d3.selectAll('.highlight')
+        .classed('highlight', false);
+    d3.select("#rank-data-text").select('text')
+        .text(" ");    
+      rankSvg.updateStyles();  
       
-      d3.select(this).attr("transform", "translate(" + (d3.event.x) + "," + (d3.event.y) + ")");
   }
-  function overRankBar(shapeBox){
   
-    var openBars = d3.selectAll('.open-rank-bar').nodes();
-    for (var i = 0; i < openBars.length; i++) { 
-      var thisBar = openBars[i];
-      var barBox = thisBar.getBoundingClientRect();
+  lockedBars
+    .style('fill', categoryToColor('total'))
+    .style('stroke-width', 0)
+    .on('mouseover',function(){
+      var state = d3.select(this).attr('id').split('-')[0];
+      d3.select('#'+state+'-locked').classed('highlight', true);
+      d3.select(this).classed('highlight', true);
+      rankSvg.updateStyles();
+      d3.select("#rank-data-text").select('text')
+        .text(this.data().abrv+": "+this.data().wu);
+    })
+    .on('mouseout',clearHighlight);
+  
     
-      var isOverBar = !(barBox.right < shapeBox.left || 
-                barBox.left > shapeBox.right || 
-                barBox.bottom < shapeBox.top || 
-                barBox.top > shapeBox.bottom);
-      if (isOverBar){
-        d3.select(thisBar).classed('chosen-rank-bar',true);
-      } else {
-        d3.select(thisBar).classed('chosen-rank-bar',false);
-      }
-    }
+  openBars
+    .style('fill', "rgba(255, 255, 255, 0.0)")
+    .style('stroke',"rgb(190,190,190)")
+    .style('stroke-width', 1.5)
+    .style("stroke-dasharray","4, 2")
+    .on('mouseover',function(){
+      d3.select(this).classed('highlight', true);
+      d3.select("#rank-data-text").select('text')
+        .text(this.data().wu);
+      rankSvg.updateStyles();
+    })
+    .on('mouseout',clearHighlight);
+    
+  lockedStates
+    .style('fill',"rgb(220,220,220)")
+    .style('stroke',"rgb(190,190,190)")
+    .style('stroke-width', 4)
+    .style("stroke-dasharray","10, 10")
```

Comment:
In other situations I've tried to push any style info I could into css, just switching the class in a call like this one, or into styles.js if required to support animated transitions among styles. My goals have been (1) consolidating styles into 1 or 2 files so it's easy for others to edit them and (2) possibly?? speeding up the code by doing less in d3, more with css. But I don't actually know if 2 is true (could be negligible or maybe even a slowdown), and 1 doesn't feel like a big deal to me especially as more of us become comfortable reading and tweaking d3 code.

So I definitely don't think we need to change this code, but I'm curious about why you chose to put the styling here this time around...maybe if we keep discussing reasoning we'll eventually come up with a unified philosophy on when to put styling where.

Reply:
I ran into some part of this that I had to do in js, so I just slid everything into js so it didn't spill over into both. I think it was the dash-array part that just didn't work well in css-only.

Reply:
:heavy_check_mark: makes sense